File: Reqnroll/Chapter20/20.4-ParameterizedScenarioPattern/Before/WIMP.Specs/StepDefinitions/PricingStepDefinitions.cs
using Reqnroll;

using WIMP.App.Models;
using WIMP.App.Services;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class PricingStepDefinitions(PricingService pricingService, IReqnrollOutputHelper outputHelper)
{
    private readonly Dictionary<string, CurrencyValue> pizzaNetPrices = new();
    private OrderItemPrice? calculatedPrice;

    [Given("the net price of the {word} pizza is ${decimal}")]
    public void GivenTheNetPriceOfThePizzaIs(string pizzaName, decimal netUnitPrice)
    {
        pizzaNetPrices[pizzaName] = new CurrencyValue(netUnitPrice, "USD");
    }

    [When("the price of {int} {word} pizzas are calculated")]
    public void WhenThePriceOfPizzasAreCalculated(int count, string pizzaName)
    {
        var netUnitPrice = pizzaNetPrices[pizzaName];
        calculatedPrice = pricingService.CalculateItemPrice(netUnitPrice, count);
        outputHelper.WriteLine($"Calculated price: {calculatedPrice}");
    }

    [Then("the sum net price is ${decimal}")]
    public void ThenTheSumNetPriceIs(decimal expectedPrice)
    {
        Assert.AreEqual(expectedPrice, calculatedPrice?.NetPrice.Value);
    }

    [Then("the sales tax is {decimal} percent")]
    public void ThenTheSalesTaxIsPercent(decimal expectedSalesTaxPercent)
    {
        Assert.AreEqual(expectedSalesTaxPercent, calculatedPrice?.SalesTaxPercent);
    }
}

File: CucumberJVM/Chapter20/20.4-ParameterizedScenarioPattern/Before/src/test/java/com/wimp/app/specs/stepdefinitions/PricingStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.models.CurrencyValue;
import com.wimp.app.models.OrderItemPrice;
import com.wimp.app.services.PricingService;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

public class PricingStepDefinitions {
    private final PricingService pricingService;
    private final Map<String, CurrencyValue> pizzaNetPrices = new HashMap<>();
    private OrderItemPrice calculatedPrice;

    public PricingStepDefinitions(PricingService pricingService) { this.pricingService = pricingService; }

    @Given("the net price of the {word} pizza is ${decimal}")
    public void theNetPriceOfThePizzaIs(String pizzaName, BigDecimal netUnitPrice) {
        pizzaNetPrices.put(pizzaName, new CurrencyValue(netUnitPrice, "USD"));
    }

    @When("the price of {int} {word} pizzas are calculated")
    public void thePriceOfPizzasAreCalculated(int count, String pizzaName) {
        calculatedPrice = pricingService.calculateItemPrice(pizzaNetPrices.get(pizzaName), count);
    }

    @Then("the sum net price is ${decimal}")
    public void theSumNetPriceIs(BigDecimal expectedPrice) {
        assertThat(calculatedPrice == null ? null : calculatedPrice.netPrice()).isEqualByComparingTo(new CurrencyValue(expectedPrice, "USD"));
    }

    @Then("the sales tax is {decimal} percent")
    public void theSalesTaxIsPercent(BigDecimal expectedPercent) {
        assertThat(calculatedPrice == null ? null : calculatedPrice.salesTaxPercent()).isEqualByComparingTo(expectedPercent);
    }
}
